[coreml] delegate multiple outputs

mcr229 opened the pull request:
Summary:
https://www.internalfb.com/code/fbsource/[c0e4da0b5c7fff3b4e31e4611033c30cabdc6aef]/fbcode/caffe2/torch/csrc/jit/backends/backend_detail.cpp?lines=268-276

seems like the torchscript addition of
`$unpack, = self.__backend.execute( ... `

the comma after unpack forces the result of execute to have only one item. So for this fix now when the size of the outputs > 1, execute returns a List List of outputs (basically put the outputs in another list before putting it into the list we return)
```
[[output1, output2, output3, ...]]
```
instead of
```
[output1, output2, output3, ...]
```

Do we want to fix this in backend_detail? Or should we make the change in our delegate to accomadate the torchscript? Proposing this q here. Requesting cccclai, kimishpatel for approval here

Test Plan: unblocked models for chengxiangyin and models in pytorch playground all passing unit tests

Reviewed By: kimishpatel, cccclai

Differential Revision: D40328684

Skylion007 commented on torch/csrc/jit/backends/coreml/objc/PTMCoreMLBackend.mm:
```diff
@@ -90,6 +90,11 @@ GenericList pack_outputs(const std::vector<TensorSpec>& output_specs, id<MLFeatu
   }
+  if(output_specs.size() > 1){
+    c10::List<c10::List<torch::Tensor>> output_res;
+    ouptut_res.push_back(outputs);
```
```suggestion
    output_res.push_back(std::move(outputs));
```

Skylion007 reviewed (commented):
One other nit

Skylion007 commented on torch/csrc/jit/backends/coreml/objc/PTMCoreMLBackend.mm:
```diff
@@ -90,6 +90,11 @@ GenericList pack_outputs(const std::vector<TensorSpec>& output_specs, id<MLFeatu
       count * sizeof(float));
     outputs.push_back(tensor);
```
```suggestion
    outputs.push_back(std::move(tensor));
```

Skylion007 replied:
Also the std::vector above it should be call .reserve() to reserve the length since the length is known.

Skylion007 reviewed (commented):
Also this one.

Skylion007 reviewed (commented):
Looks good, this piece of code could definitely be optimized a lot with more std::move and reserve() etc... for all the push_back, toList etc...

Skylion007 commented on torch/csrc/jit/backends/coreml/objc/PTMCoreMLBackend.mm:
```diff
@@ -90,6 +90,11 @@ GenericList pack_outputs(const std::vector<TensorSpec>& output_specs, id<MLFeatu
+  if(output_specs.size() > 1){
+    c10::List<c10::List<torch::Tensor>> output_res;
+    output_res.push_back(std::move(outputs));
+    return c10::impl::toList(output_res);
+  }
   return c10::impl::toList(outputs);
```
Should also move here actually.
```suggestion
    return c10::impl::toList(std::move(output_res));
  }
  return c10::impl::toList(std::move(outputs));
```

Skylion007 commented on torch/csrc/jit/backends/coreml/objc/PTMCoreMLBackend.mm:
```diff
@@ -90,6 +90,11 @@ GenericList pack_outputs(const std::vector<TensorSpec>& output_specs, id<MLFeatu
       count * sizeof(float));
     outputs.push_back(tensor);
```
```suggestion
    outputs.push_back(std::move(tensor));
```

Skylion007 replied:
Also see the std::vector above and properly call .reserve() on it etc.

mcr229 committed:
[coreml] delegate multiple outputs (#88345)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/88345

https://www.internalfb.com/code/fbsource/[c0e4da0b5c7fff3b4e31e4611033c30cabdc6aef]/fbcode/caffe2/torch/csrc/jit/backends/backend_detail.cpp?lines=268-276

seems like the torchscript addition of
`$unpack, = self.__backend.execute( ... `

the comma after unpack forces the result of execute to have only one item. So for this fix now when the size of the outputs > 1, execute returns a List List of outputs (basically put the outputs in another list before putting it into the list we return)
```
[[output1, output2, output3, ...]]
```
instead of
```
[output1, output2, output3, ...]
```

Do we want to fix this in backend_detail? Or should we make the change in our delegate to accomadate the torchscript? Proposing this q here. Requesting cccclai, kimishpatel for approval here

Test Plan: unblocked models for chengxiangyin and models in pytorch playground all passing unit tests

Reviewed By: kimishpatel, cccclai

Differential Revision: D40328684

fbshipit-source-id: bfb85f3f0933b062a14c5d8cd3acb7db58c52944

Skylion007 reviewed (approved):
In the changed function, the std::vector `output_shape` should also allocate it's space using `.reserve()` before the for loop. Otherwise this code is looking much better and more performant.
